src/gui/torrentcategorydialog.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2017  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QDialog>
+
+namespace Ui
+{
+    class TorrentCategoryDialog;
+}
+
+class TorrentCategoryDialog: public QDialog
+{
+    Q_OBJECT
+    Q_DISABLE_COPY(TorrentCategoryDialog)
+
+public:
+    static QString createCategory(QWidget *parent, const QString &parentCategoryName = "");
+    static void editCategory(QWidget *parent, const QString &categoryName);
+
+    explicit TorrentCategoryDialog(QWidget *parent = nullptr);
+    ~TorrentCategoryDialog() override;
+
+    void setCategoryNameEditable(bool editable);
```

Comment:
I would merge this one into `setCategoryName(const QString&, bool)`

Reply:
I would prefer to leave it as is. 
There is nothing illegal. In another use case we could, for example, to have a single instance and reuse it, changing the properties as needed before call `exec()`.

Reply:
OK, but the constructor now creates an "incomplete" object. Might make sense to add these parameters to the constructor and call the current functions from it.

Reply:
>but the constructor now creates an "incomplete" object. 

Why "incomplete"? It's just "default".
This design (used in most places in Qt) allows to avoid monstrous constructor in case the object can be fine tuned using many properties (now we have only name and save path but it is created with an eye to the future, because we already have a lot of requests to do many different "auto-managed" properties).

Reply:
Incomplete because it does not have category name set, and is nonoperational as such. 

Why don't you refer to the more modern part of the Qt, the one which uses models? The dialog could be implemented in that way, taking a category object (string-string pair as of now) as the model, and modifying it? This would be so less intrusive and much more easily extendable later on.

Reply:
Well, how do you propose to distinguish between what the current purpose the dialog created? Creating new category or editing the existing one? First case requires parent category to be given but the second the edited category itself.

Reply:
Actually I like model-based approach. But I have absolutely no time to change that now. If it is not critical, I would prefer to return to this issue when implementing the TorrentCategory class. Otherwise, this PR will hang indefinitely (currently I'm too busy on my main job).

Reply:
>Why don't you refer to the more modern part of the Qt, the one which uses models?

Well, modern is not always best... But simple things always work.
Now it's just a specific dialog with some controls, we can set some initial values, the user can change anything, then we can get existing values and use them. It's a universal way. Next is another layer that adds a little convenience (static methods createCategory/editCategory).
Model-based way is modern and powerfull but it requires the model should exist. We need both edit an existing Category and create a new one using similar way (the same UI). Of course, we can split creating Category into creating itself and editing newly created category, but, IMO, it isn't good idea in this case.
@evsh, do you have definite suggestions?

Reply:
My suggestion is as follows. Instead of all get/set methods, create the single `void setCategory(CategoryObject*, EditingFlags)` function. Specify editing/creating mode with flags. Add more flags to make certain category properties read-only when needed. I do not insists, of course.

Reply:
But I don't understand why you want to pull it below! This is implemented on a higher level (as I described in the post above). I mean those two static methods: createCategory and editCategory. When I implement the TorrentCategory class, it will look like this:
```c++
static TorrentCategory *createCategory(QWidget *parent, TorrentCategory *parentCategory);
static void editCategory(QWidget *parent, TorrentCategory *category);
```
>I do not insists, of course.

In any case, I don't want to do this now in the absence of the affected class. Perhaps my opinion will change when I'll work on it.